harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +84,356 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+
+        .. code-block:: python
+
+        class MyModel:
+            def __init__(self):
+                self.a = A(...)
+                self.b = B(...)
+
+            def forward(self):
+                ...
+
+        self.context = trial_context
+        self.model = self.context.Model(MyModel())
+        """
+
+        check.is_none(
+            self.model,
+            f"Must call Model() only once. See the instruction below: {self._Model.__doc__}",
+        )
+        self.model = model.to(self.device)
+        if not self.hvd_config.use and self.n_gpus > 1:
+            check.eq(
+                self.hvd_config.aggregation_frequency,
+                1,
+                "Please enable `optimized_parallel` to use aggregation "
+                "frequency greater than 1 for single machine multi-GPU "
+                "training.",
+            )
+            self.model = nn.DataParallel(self.model)
+            logging.debug("Initialized mode for native parallel training.")
+        return model
+
+    def _Optimizer(self, optimizer: torch.optim.Optimizer) -> torch.optim.Optimizer:  # type: ignore
+        """Optimizer() initializes an optimizer. It returns an initialized optimizer.
+
+        The optimizer must use a model initialized by Model() because
+        torch.optim.Optimizer directly access the model parameters,
+        which are initialized in Model().
+
+        Optimizer creates a horovod.DistributedOptimizer if uses parallel/distributed training.
+        """
+
+        # TODO: Check that optimizer is not an amp optimizer.
+        if self.hvd_config.use:
+
+            def filter_named_parameters(
+                model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+            ) -> List:
+                """
+                filter_named_parameters filters the named parameters of a specified optimizer out
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
```

Comment:
This is unpleasant enough that I think we should weight the pros and cons of the API against what we have to do behind the scenes.

So the reason for this, as @shiyuann explained to me, is that horovod needs the named parameters.  Horovod expects the user to feed the `named_parameters` into the horovod optimizer, and presumably they need the to match the `parameters` that were fed to the optimizer:

```python
optimizer = optim.SGD(model.parameters())
optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
```

The fact is that you can go from named_parameters to parameters, but the reverse lookup is not natively supported.

So I see three options:

1. do a reverse lookup of the named_parameters from the parameters based on our list of all the models (the current PR implementation)
2. mimic the horovod API and make users hand us the named_parameters explicitly
3. we hook into the optimizer initialization and users hand us named_parameters

### 1. do a reverse lookup

(code example is just basically the current implementation)

pros:

- API is super easy to use

cons:

- It's not clear to me if there are use cases that we are going to be unable to support.  I imagine that if the parameters in an optimizer were not already part of one of the registered models, something else would probably break anyway, so this might not be so bad.

### 2. minimc the horovod API

```python
optimizer = optim.SGD(model.parameters())
optimizer = context._Optimizer(optimizer, named_parameters=model.named_parameters())
```

pros:

- Solution is pretty bulletproof; if horovod supports the training then we would too, because we're essentially passing the horovod API straight to the user.

cons:

- The API is harder to use, and users who don't care about dtrain still need to adjust their code to fit a dtrain paradigm

### 3. hook into the optimizer initialization

```python
# under the hood, we instantiate the optimize by stripping names from the named_parameters
optimizer = context._make_optimizer(optim.SGD, model.named_parameters())
```

pros:

- Literally every `Optimizer` in `torch` takes parameters as its first `__init__` parameter, so this API would be reasonably natural.
- No reverse lookup necessary

cons:

- This would kind of create a weird API that I wouldn't love if I was a user.  I think wrapping constructed objects is more natural than wrapping constructed.

### Conclusion

I actually changed my own mind while writing this.  I started thinking that option 3 was the best but now I think that option 1 (what you already have) is the best, unless anyone (@aaron276h, @katport) can think of a case where the parameters of an optimizer are not part of one of the models (since that would break the reverse lookup).

@shiyuann  still, why don't we simplify the reverse lookup by building a dictionary from all of the models we have registered in the context (see below)?  It seems like this would be a very clean mechanism and wouldn't require users to build all of their models as a single model with submodels.

```python
class PyTorchContext(TrainContext):
     def __init__(self, ...):
          self._parameter_names = {}

     def _Model(self, model):
          ...
          for name, param in model.named_parameters():
              self._parameter_names[parameter] = name
     
     def _Optimizer(self, opt):
          ...
          named_parameters = {self._named_parameters[p] for p in opt_params}
          optimzer = hvd.DistributedOptimizer(
               opt,
               named_parameters = named_parameters,
               ...
           )
           ...
```

Reply:
>  @shiyuann still, why don't we simplify the reverse lookup by building a dictionary from all of the models we have registered in the context (see below)? It seems like this would be a very clean mechanism and wouldn't require users to build all of their models as a single model with submodels.

Because `named_parameters()` is way more complex than simply just adding parameters to a dict. You need to consider the cases if there are sub-modules refer to the same module. Basically, we'll have to duplicate the logic in `nn.Module.named_modules()`.

Reply:
If we decide to go down the path of supporting multiple models (as I understand the current agreed on plan required users to submit one model that they have multiple modules inside of) I think the most transparent approach is requiring named_parameters (approach #2).

Reply:
@shiyuann, I think we need to understand what Horovod needs the named_parameters for in the first place.

My gut sense was pytorch optimizers keep a list of parameters for tracking gradient updates (no name needed since they can just keep the objects in memory), and the only reason that horovod needs named parameters is to know which parameters in one process' memory space correspond with parameters in the other proccess' memory space.  This turns out to be true (https://github.com/horovod/horovod/blob/master/horovod/torch/optimizer.py#L32)

Horovod already handles the case where multiple names point to a single parameter; they just sort the list of named parameters by name when they build their reverse-lookup dictionary (which they actually call `self.parameter_names` too haha): https://github.com/horovod/horovod/blob/master/horovod/torch/optimizer.py#L64.  Admittedly, the solution I proposed isn't quite right, but it's not very far from right either.

@aaron276h I agree that option 2 is more transparent, but is there any possible situation where a user would have a trainable tensor that is not part of one of the models that they register with us, but where we still support their use case?  I think we can safely hide this named_parameter details from users since we already require users to register every model in their system with us.  This is different than native Horovod, since native Horovod only actually hooks into the optimizer.

@shiyuann, @aaron276h There's really two questions here:

A. Is there a technical reason why we need to enforce the single-model-with-submodels strategy on users?  (this was the original motivation for this investigation)
B. What should our `context._Optimizer()` api look like?

Regarding A, I don't see any technical reason why we need to enforce the single-model-with-submodels strategy on users.  I think that would be a pretty arbitrary requirement of our API.

Regarding B, @aaron276h, does what I said here change your mind?

Reply:
< 
@aaron276h I agree that option 2 is more transparent, but is there any possible situation where a user would have a trainable tensor that is not part of one of the models that they register with us, but where we still support their use case? I think we can safely hide this named_parameter details from users since we already require users to register every model in their system with us. This is different than native Horovod, since native Horovod only actually hooks into the optimizer.

I think we can safely assume all the parameters that need to update are registered in at least one module.

Reply:
For question A, as I said in other threads, it's not blocking but it makes the implementation unnecessarily complex. 

It will be:

```
params_a = model_a.named_parameters()
params_b = model_a.named_parameters()
all_params = {}
# remove duplicate module
params = set()
prefix="params_a"
for k, v in params_a:
    if v in params:
         pass
    params.add(v)
    params[prefix+k] = v

# Do the same for all the modules.
...
# Find the parameters of the optimizer in all the parameters.
...
```

Reply:
Update after discussing it offline: we will go with multiple models